tests/components/heos/test_media_player.py

```diff
@@ -805,3 +810,93 @@ async def test_play_media_invalid_type(hass, config_entry, config, controller, c
         blocking=True,
     )
     assert "Unable to play media: Unsupported media type 'Other'" in caplog.text
+
+
+async def test_media_player_join_group(hass, config_entry, config, controller, caplog):
+    """Test grouping of media players through the join service."""
+    await setup_platform(hass, config_entry, config)
+    await hass.services.async_call(
+        MEDIA_PLAYER_DOMAIN,
+        SERVICE_JOIN,
+        {
+            ATTR_ENTITY_ID: "media_player.test_player",
+            ATTR_GROUP_MEMBERS: ["media_player.test_player_2"],
+        },
+        blocking=True,
+    )
+    controller.create_group.assert_called_once_with(
+        1,
+        [
+            2,
+        ],
+    )
+    assert "Failed to group media_player.test_player with" not in caplog.text
+
+    controller.create_group.side_effect = HeosError("error")
+    await hass.services.async_call(
+        MEDIA_PLAYER_DOMAIN,
+        SERVICE_JOIN,
+        {
+            ATTR_ENTITY_ID: "media_player.test_player",
+            ATTR_GROUP_MEMBERS: ["media_player.test_player_2"],
+        },
+        blocking=True,
+    )
+    assert "Failed to group media_player.test_player with" in caplog.text
+
+
+async def test_media_player_group_members(
+    hass, config_entry, config, controller, caplog
+):
+    """Test group_members attribute."""
+    await setup_platform(hass, config_entry, config)
+    hass.data[DOMAIN][DATA_GROUP_MANAGER].force_update_groups()
```

Comment:
Please don't access integration details in the test. Set up the integration and let it add the entity and update the group.

Reply:
It turns out that after my last refactoring `force_update_groups ` doesn't have to be called at all during tests, as the player instances handle this. :smile: